### PR TITLE
feat(feature-flags): Enable experience continuity with feature flags

### PR DESCRIPTION
**Pull request by neilkakkar**
## Problem

Have a look here: https://github.com/PostHog/posthog/issues/9547 for context. We're going with option 3!

The code in `posthog/models/feature_flag.py` might seem a bit wrangly, but there's a good reason for it: My goal here was to make sure:

1. When there are no special cased feature flags (experience_continuity enabled), there's no slowdown. `decide` endpoint takes exactly the same number of queries as before.
2. When there are special cased feature flags, the increase in querying is minimal, and the response is sufficiently quick.

There's some cases we still don't handle well, specially to do with ingestion delays, but I don't think there's a simple way around that. We're borked anyway in that case, since person properties filtering won't work, which means the feature flag, even if it had experience continuity, wouldn't have worked.

## Changes

👉 *Stay up-to-date with [PostHog coding conventions](https://posthog.com/docs/contribute/coding-conventions) for a smoother review.*

### CODE REVIEW

File: frontend/src/scenes/feature-flags/FeatureFlag.tsx [hunk 1]
```diff
@@ -179,6 +179,22 @@ export function FeatureFlag({ id }: { id?: string } = {}): JSX.Element {
                                     placeholder="Adding a helpful description can ensure others know what this feature is for."
                                 />
                             </Field>
+
```
Comment (neilkakkar) [inline frontend/src/scenes/feature-flags/FeatureFlag.tsx:182]: move to another pr!

**Comment by clarkus** [comment]
I put together an idea in figma for how we can communicate this functionality to users. https://www.figma.com/file/gQBj9YnNgD8YW4nBwCVLZf?node-id=5400:35869#208945086

Screen Shot 2022-06-09 at 12 12 19 PM

File: posthog/test/__snapshots__/test_feature_flag.ambr [hunk 2]
```diff
@@ -1,10 +1,121 @@
...
+  '
+---
+# name: TestFeatureFlagHashKeyOverrides.test_entire_flow_with_hash_key_override.6
+  '
+  SELECT "posthog_featureflagoverride"."id",
+         "posthog_featureflagoverride"."feature_flag_id",
+         "posthog_featureflagoverride"."override_value",
+         "posthog_featureflag"."id",
+         "posthog_featureflag"."key"
+  FROM "posthog_featureflagoverride"
```
Comment (neilkakkar) [inline posthog/test/__snapshots__/test_feature_flag.ambr:68]: this cruft from existing overrides makes me sad.

Specially, since this applies to a very small minority of users: Those who have a posthog account, while being called for every user coming to the client's website.

At minimum, we should just check if this distinct ID belongs to a posthog user, and then use the value inside this query. Hmm, but then, if featureflagoverride table is small, not sure it makes a difference.

**Comment by neilkakkar** [comment]
Asking for ingestion side reviews as well, since there's an update when merging two persons together. I think it's fine to do as I have, since we're in a transaction, but lmk if I'm missing something, @tiina303 , @macobo .

Edit: Also, If I'm not mistaken, the deploy will ensure that migrations run before django & plugin server updates, right? Or do I need to split out the plugin-server part to a separate PR?

File: plugin-server/tests/main/process-event.test.ts [hunk 3]
```diff
@@ -1475,6 +1476,184 @@ test('distinct with anonymous_id which was already created', async () => {
     expect(person.is_identified).toEqual(true)
 })
 
+test('distinct with anonymous_id which was already created triggers foreign key updates', async () => {
```
Comment (neilkakkar) [inline plugin-server/tests/main/process-event.test.ts:1479]: hmm, should add a test in django-land for where old user is identified, which means we won't reach this code path, and no updates would occur. Will the override django code handle that?

File: plugin-server/src/worker/ingestion/process-event.ts [hunk 4]
```diff
@@ -798,4 +792,62 @@ export class EventsProcessor {
...
+        await this.db.postgresQuery(
+            'UPDATE posthog_cohortpeople SET person_id = $1 WHERE person_id = $2',
+            [targetPerson.id, sourcePerson.id],
+            'updateCohortPeople',
+            client
+        )
+
+        // For FeatureFlagHashKeyOverrides
+        const allOverrides = await this.db.postgresQuery(
+            'SELECT id, person_id, feature_flag_key FROM posthog_featureflaghashkeyoverride WHERE team_id = $1 AND person_id = ANY($2)',
```
Comment (macobo) [inline plugin-server/src/worker/ingestion/process-event.ts:813]: (didn't look further into this PR): Why do we select + do post-processing + update instead of a single update query? We're introducing a lot of new failure modes here.

File: posthog/api/decide.py [hunk 5]
```diff
@@ -150,7 +150,9 @@ def get_decide(request: HttpRequest):
             team = user.teams.get(id=project_id)
 
         if team:
-            feature_flags = get_overridden_feature_flags(team.pk, data["distinct_id"], data.get("groups", {}))
+            feature_flags = get_overridden_feature_flags(
```
Comment (macobo) [inline posthog/api/decide.py:153]: If we're touching /decide endpoint, that should be a separate PR to ensure easy control if something goes wrong. This single endpoint breaking can bring down all of posthog-js ingestion in nasty ways.

File: plugin-server/src/worker/ingestion/process-event.ts [hunk 6]
```diff
@@ -798,4 +792,62 @@ export class EventsProcessor {
             )
         }
     }
+
+    private async handleTablesDependingOnPersonID(
```
Comment (macobo) [inline plugin-server/src/worker/ingestion/process-event.ts:796]: Note: This should be a separate PR given the risk of accidentally breaking ingestion.

File: plugin-server/src/worker/ingestion/process-event.ts (hunk 4 above)
Comment (macobo) [reply plugin-server/src/worker/ingestion/process-event.ts:813]: A https://stackoverflow.com/questions/5874453/postgresql-way-to-insert-row-with-on-conflict-clause-semantics might work better if you're looking for UPDATE but avoids conflicts case. - just add `ON CONFLICT DO NOTHING` to the INSERT.

File: plugin-server/tests/main/process-event.test.ts [hunk 7]
```diff
@@ -41,7 +42,7 @@ export async function createPerson(
         {},
         team.id,
         null,
-        false,
+        identified,
```
Comment (macobo) [inline plugin-server/tests/main/process-event.test.ts:45]: Q: Where are we using `is_identified` in code?

File: posthog/models/feature_flag.py [hunk 8]
```diff
@@ -333,3 +421,26 @@ def get_overridden_feature_flags(
...
     return feature_flags
+
+
+def set_feature_flag_hash_key_overrides(
+    feature_flags: QuerySet, team_id: int, person_id: int, hash_key_override: str
+) -> None:
+
+    existing_flag_overrides = set(
+        val.feature_flag_key
+        for val in FeatureFlagHashKeyOverride.objects.filter(team_id=team_id, person_id=person_id).only(
```
Comment (macobo) [inline posthog/models/feature_flag.py:432]: You can use `.values_list("feature_flag_key", flat=True)` and remove the list comprehension

File: posthog/models/feature_flag.py [hunk 9]
```diff
@@ -311,12 +356,55 @@ def get_active_feature_flags(
...
+            # the new person on which decide was called.
+            # In this case, we can try finding the person_id for the old distinct id.
+            # This is safe, since once `$identify` is processed, it would only add the distinct_id to this
+            # existing person. If, because of race conditions, a person merge is called for later,
+            # then https://github.com/PostHog/posthog/blob/master/plugin-server/src/worker/ingestion/process-event.ts#L480
+            # will take care of it^.
+            try:
+                person_id = PersonDistinctId.objects.filter(distinct_id=hash_key_override, team_id=team_id).values_list(
+                    "person_id"
+                )[:1][0][0]
```
Comment (macobo) [inline posthog/models/feature_flag.py:389]: This is super crufty code. Do we need the [:1], could we use `flat=True` for values_list?

File: posthog/models/feature_flag.py [hunk 10]
```diff
@@ -311,12 +356,55 @@ def get_active_feature_flags(
 
 # Return feature flags with per-user overrides
 def get_overridden_feature_flags(
-    team_id: int, distinct_id: str, groups: Dict[GroupTypeName, str] = {},
+    team_id: int, distinct_id: str, groups: Dict[GroupTypeName, str] = {}, hash_key_override: Optional[str] = None
 ) -> Dict[str, Union[bool, str, None]]:
-    feature_flags = get_active_feature_flags(team_id, distinct_id, groups)
```
Comment (macobo) [inline posthog/models/feature_flag.py:316]: Is this feature flag still used anywhere? If so, it's a sign that maybe it shouldn't be?

File: posthog/test/__snapshots__/test_feature_flag.ambr (hunk 2 above)
Comment (macobo) [reply posthog/test/__snapshots__/test_feature_flag.ambr:68]: Could we kill this then? This won't affect any real operations and we could if we wanted to.

File: posthog/models/feature_flag.py (hunk 10 above)
Comment (neilkakkar) [reply posthog/models/feature_flag.py:316]: not sure I get this. Do you mean overrides only applying to active feature flags? Or something else?

File: posthog/models/feature_flag.py [hunk 11]
```diff
@@ -289,19 +318,35 @@ def _variant_hash(self) -> float:
         return self.get_hash(salt="variant")
 
 
+def hash_key_overrides(team_id: int, person_id: int) -> Dict[str, str]:
+    feature_flag_to_key_overrides = {}
+    for feature_flag, override in FeatureFlagHashKeyOverride.objects.filter(
+        person_id=person_id, team=team_id
```
[the remark this answers is not in the document]
Comment (neilkakkar) [reply posthog/models/feature_flag.py:324]: the unique index is team_id+person_id+feature_flag_key, so should work, but will confirm

File: posthog/models/feature_flag.py (hunk 10 above)
Comment (macobo) [reply posthog/models/feature_flag.py:316]: I mean `get_active_feature_flags` as a function

File: posthog/models/feature_flag.py [hunk 12]
```diff
@@ -289,19 +318,35 @@ def _variant_hash(self) -> float:
...
+    return feature_flag_to_key_overrides
+
+
 # Return a Dict with all active flags and their values
 def get_active_feature_flags(
-    team_id: int, distinct_id: str, groups: Dict[GroupTypeName, str] = {},
+    feature_flags: QuerySet,
+    team_id: int,
+    distinct_id: str,
+    person_id: Optional[int] = None,
```
[the remark this answers is not in the document]
Comment (neilkakkar) [reply posthog/models/feature_flag.py:336]: my thinking: If postgres is source of truth, I don't care what's in clickhouse.

And in postgres, the id is the pk, and used wherever a foreign key relation is needed (`CohortPeople` )

[the remark this answers is not in the document]
Comment (macobo) [reply posthog/models/feature_flag.py:336]: Fair enough!

File: posthog/test/__snapshots__/test_feature_flag.ambr (hunk 2 above)
Comment (neilkakkar) [reply posthog/test/__snapshots__/test_feature_flag.ambr:68]: Good question. I think killing this will remove the toolbar feature flags functionality. Don't think it's used anywhere else, is it? (cc: @mariusandra @rcmarron )

At the very least, I suppose we could go down this code path only when asking for `/my_flags` or `/decide` has a parameter used by the toolbar to make it happen.

This way, we can pretty much get the best of both worlds, by keeping this functionality, and removing cruft completely for most users.

Although I do question if the toolbar override functionality is used at all, @marcushyett-ph ?

File: posthog/models/feature_flag.py (hunk 11 above)
[the remark this answers is not in the document]
Comment (neilkakkar) [reply posthog/models/feature_flag.py:324]: it does

File: posthog/models/feature_flag.py [hunk 13]
```diff
@@ -311,12 +356,55 @@ def get_active_feature_flags(
...
+
     person = PersonDistinctId.objects.filter(distinct_id=distinct_id, team_id=team_id).values_list("person_id")[:1]
+
+    if flags_have_experience_continuity_enabled:
+        try:
+            person_id = person[0][0]
+        except IndexError:
+            person_id = None
+
+        if person_id is None and hash_key_override is not None:
```
Comment (EDsCODE) [inline posthog/models/feature_flag.py:378]: NIT: Can we move `if hash_key_override` to a top level if and make the person_id if/else more obvious. Was going back and forth when looking at this sequence of checks

File: posthog/models/feature_flag.py [hunk 14]
```diff
@@ -311,12 +356,55 @@ def get_active_feature_flags(
...
+            # :TRICKY: Some ingestion delays may mean that `$identify` hasn't yet created
+            # the new person on which decide was called.
+            # In this case, we can try finding the person_id for the old distinct id.
+            # This is safe, since once `$identify` is processed, it would only add the distinct_id to this
+            # existing person. If, because of race conditions, a person merge is called for later,
+            # then https://github.com/PostHog/posthog/blob/master/plugin-server/src/worker/ingestion/process-event.ts#L480
+            # will take care of it^.
+            try:
```
Comment (EDsCODE) [inline posthog/models/feature_flag.py:386]: I think you can use `.first()` here 

Ref: https://docs.djangoproject.com/en/dev/ref/models/querysets/#first

File: posthog/models/feature_flag.py [hunk 15]
```diff
@@ -167,11 +188,15 @@ def __init__(
         distinct_id: str,
         groups: Dict[GroupTypeName, str] = {},
         cache: Optional[FlagsMatcherCache] = None,
+        person_id: Optional[str] = None,
+        hash_key_overrides: Dict[str, str] = {},
     ):
         self.feature_flag = feature_flag
         self.distinct_id = distinct_id
+        self.person_id = person_id
```
Comment (EDsCODE) [inline posthog/models/feature_flag.py:196]: Where is this used?

**Review by EDsCODE — APPROVED** [review]
Note: I have not worked with the flag matching logic in a while so I may not be 100% aware of the potholes in all the related areas that are being affected here. But, rolling out according to @macobo's should help a lot

Good work!

File: posthog/models/feature_flag.py (hunk 15 above)
Comment (neilkakkar) [reply posthog/models/feature_flag.py:196]: Nowhere anymore. Will remove!
